sympy/core/function.py

```diff
@@ -1244,33 +1244,17 @@ def __new__(cls, expr, *variables, **assumptions):
                 obj = None
             else:
                 if isinstance(v, (collections.Iterable, Tuple, MatrixCommon, NDimArray)):
-                    deriv_fun = derive_by_array
                     is_symbol = True
-                else:
-                    deriv_fun = lambda x, y: x._eval_derivative(y)
                 if not is_symbol:
                     new_v = Dummy('xi_%i' % i, dummy_index=hash(v))
                     expr = expr.xreplace({v: new_v})
                     old_v = v
                     v = new_v
-                obj = expr
                 if count.is_Integer:
-                    for i in range(count):
-                        obj2 = deriv_fun(obj, v)
-                        if obj == obj2:
-                            break
-                        obj = obj2
-                        nderivs += 1
-                elif obj.is_Derivative:
-                    dict_var_count = dict(obj.variable_count)
-                    if v in dict_var_count:
-                        dict_var_count[v] += count
-                    else:
-                        dict_var_count[v] = count
-                    obj = Derivative(obj.expr, *dict_var_count.items())
-                    nderivs += count
+                    obj = Derivative._helper_apply_n_times(expr, v, count, lambda x, y: x._eval_derivative(y))
```

Comment:
This check should be removed. If an object wants to manually fallback to `_eval_derivative`, it can do that, but I think `_eval_derivative_n_times` should always be preferred if it is defined (the default fallback will happen automatically due to the definition on Basic).

Reply:
We need to handle expressions of arrays as well, which is making the API a tad more complicated.